internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -215,9 +224,132 @@ func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path)
 		allErrs = append(allErrs, validateSchema(&p, fldPath.Child("properties").Key(propertyName))...)
 	}
 
+	// If any schema related validation errors have been found at this level or deeper, skip CEL expression validation.
+	// Invalid OpenAPISchemas are not always possible to convert into valid CEL DeclTypes, and can lead to CEL
+	// validation error messages that are not actionable (will go away once the schema errors are resolved) and that
+	// are difficult for CEL expression authors to understand.
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	celContext := apiextensionsvalidation.RootCELContext(schema)
+	allErrs = append(allErrs, validateCELExpressions(schema, fldPath, celContext)...)
+
+	return allErrs
+}
+
+func validateCELExpressions(schema *apiextensions.JSONSchemaProps, fldPath *field.Path, celContext *apiextensionsvalidation.CELSchemaContext) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if celContext == nil {
+		return allErrs
+	}
+
+	if schema.AdditionalProperties != nil {
+		allErrs = append(allErrs, validateCELExpressions(schema.AdditionalProperties.Schema, fldPath.Child("additionalProperties"), celContext.ChildAdditionalPropertiesContext(schema.AdditionalProperties.Schema))...)
+	}
+
+	if len(schema.Properties) != 0 {
+		for property, jsonSchema := range schema.Properties {
+			propertySchema := jsonSchema
+			allErrs = append(allErrs, validateCELExpressions(&propertySchema, fldPath.Child("properties").Key(property), celContext.ChildPropertyContext(&propertySchema, property))...)
+		}
+	}
+
 	if schema.Items != nil {
-		allErrs = append(allErrs, validateSchema(schema.Items.Schema, fldPath.Child("items"))...)
+		allErrs = append(allErrs, validateCELExpressions(schema.Items.Schema, fldPath.Child("items"), celContext.ChildItemsContext(schema.Items.Schema))...)
+		if len(schema.Items.JSONSchemas) != 0 {
+			for i, jsonSchema := range schema.Items.JSONSchemas {
+				itemsSchema := jsonSchema
+				allErrs = append(allErrs, validateCELExpressions(&itemsSchema, fldPath.Child("items").Index(i), celContext.ChildItemsContext(&itemsSchema))...)
+			}
+		}
+	}
+
+	typeInfo, err := celContext.TypeInfo()
+	if err != nil {
+		return append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), fmt.Errorf("internal error: failed to construct type information for x-kubernetes-validations rules: %w", err)))
```

Comment:
Is there a preference for wrapping errors with `%w` or `errors.Wrapf`, I've seen both while reviewing this PR

Reply:
In my opinion we should continue to use pkg/errors (errors.Wrapf) until we have consensus on: https://github.com/kubernetes-sigs/cluster-api/issues/6688#issuecomment-1911663477

Reply:
Done.